File: c/distance_range.cc
#include "distance_range.h"

#include <algorithm>
#include <limits>

#include "platform.h"
#include "sin_cos.h"

namespace twim {

DistanceRange::DistanceRange(const Vector<int32_t>& region, int32_t angle,
                             const CodecParams& cp) {
  if (region.len == 0) {
    // Exception!
    num_lines = kInvalid;
    return;
  }
  const size_t count = region.len;
  const size_t step = region.capacity / 3;
  const int32_t* RESTRICT y = region.data();
  const int32_t* RESTRICT x0 = y + step;
  const int32_t* RESTRICT x1 = x0 + step;

  // Note: kSin is all positive -> d0 <= d1
  int32_t nx = SinCos.kSin[angle];
  int32_t ny = SinCos.kCos[angle];
  int32_t mi = std::numeric_limits<int32_t>::max();
  int32_t ma = std::numeric_limits<int32_t>::min();
  for (size_t i = 0; i < count; i++) {
    int32_t d0 = ny * y[i] + nx * x0[i];
    int32_t d1 = ny * y[i] + nx * (x1[i] - 1);
    mi = std::min(mi, d0);
    ma = std::max(ma, d1);
  }
  // TODO(eustas): Check >= 0
  min = static_cast<uint32_t>(mi);
  // TODO(eustas): Check >= 0
  max = static_cast<uint32_t>(ma);

  line_quant = cp.getLineQuant();
  uint32_t delta = ma - mi;
  if (delta > line_quant * cp.line_limit) line_quant = (2 * delta + cp.line_limit) / (2 * cp.line_limit);
  num_lines = std::max<uint32_t>(delta / line_quant, 1);
}

}  // namespace twim

File: js/DistanceRange.js
import {getLineLimit, getLineQuant} from "./CodecParams.js";
import {assertFalse, b32, forEachScan} from './Mini.js';
import * as SinCos from './SinCos.js';

let /** @type{number} */ numLines = 0;
let /** @type{number} */ _min = 0;
let /** @type{number} */ _max = 0;
let /** @type{number} */ _lineQuant = 0;

/**
 * @return{number}
 */
export let getNumLines = () => numLines;

/**
 * @param{!Int32Array} region
 * @param{number} angle
 * @return {void}
 */
export let update = (region, angle) => {
  // Note: SinCos.SIN is all positive -> d0 <= d1
  let /** @type{number} */ nx = SinCos.SIN[angle];
  let /** @type{number} */ ny = SinCos.COS[angle];
  let /** @type{number} */ mi = b32;
  let /** @type{number} */ ma = -b32;
  forEachScan(region, (y, x0, x1) => {
    let /** @type{number} */ d0 = ny * y + nx * x0;
    let /** @type{number} */ d1 = ny * y + nx * x1 - nx;
    mi = mi < d0 ? mi : d0;
    ma = ma > d1 ? ma : d1;
  });
  assertFalse(ma < mi);
  _min = mi;
  _max = ma;
  _lineQuant = getLineQuant();
  let /** @type{number} */ delta = ma - mi;
  if (delta > getLineLimit() * _lineQuant) {
    _lineQuant = (2 * delta + getLineLimit()) / (2 * getLineLimit()) | 0;
  }
  numLines = delta / _lineQuant | 0;
};

/**
 * @param{number} line
 * @return {number}
 */
export let distance = (line) => {
  if (numLines > 1) {
    return _min + (((_max - _min) - (numLines - 1) * _lineQuant) / 2 | 0) + _lineQuant * line;
  } else {
    return ((_max + _min) / 2) | 0;
  }
};
